### PR TITLE
doc: add Incremental Entity Provider to main docs for better visibility

### PR DESCRIPTION
**Pull request by zeshanziya**
## Hey, I just made a Pull Request!

Add [Incremental Entity Provider](https://github.com/backstage/backstage/tree/master/plugins/catalog-backend-module-incremental-ingestion) as part of the main docs for better visibility. It was suggested [here](https://github.com/backstage/backstage/pull/26146#pullrequestreview-2254428581)

#### :heavy_check_mark: Checklist

- [ ] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [x] Added or updated documentation
- [ ] Tests for new functionality and regression tests for bug fixes
- [ ] Screenshots attached (for UI changes)
- [x] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

### CODE REVIEW

File: docs/features/software-catalog/external-integrations.md [hunk 1]
```diff
@@ -1031,3 +1031,360 @@ backend.add(catalogModuleCustomDataParser);
 
 backend.start();
 ```
+
+## Incremental Entity Provider
+
+The Incremental Ingestion catalog backend module provides an Incremental Entity Provider that can be used to ingest data from sources using delta mutations, while retaining the orphan prevention mechanism provided by full mutations.
+
+### Why did we create it?
```
Comment (freben) [inline docs/features/software-catalog/external-integrations.md:1039]: Look, this is nice text as such, but not sure it belongs here. A long motivations section, and technical implementation details, could go into the README.md of the module itself. For this article, it's probably enough to say something along the lines (paraphrasing) of "If your data source is large and perhaps does not fit into memory but does support pagination, there's a convenient incremental entity provider that lets you efficiently ingest that paginated source." And then how to install it. That's what the reader of this particular article most benefits from.

File: docs/features/software-catalog/external-integrations.md [hunk 2]
```diff
@@ -1031,3 +1031,360 @@ backend.add(catalogModuleCustomDataParser);
...
+
+/* highlight-add-start */
+backend.add(
+  import(
+    '@backstage/plugin-catalog-backend-module-incremental-ingestion/alpha'
+  ),
+);
+/* highlight-add-end */
+
+// We have created this in section **Adding an Incremental Entity Provider to the catalog**
```
Comment (freben) [inline docs/features/software-catalog/external-integrations.md:1096]:
```suggestion
// See the section below for how this provider is created.
```

a bit simpler.

There's always a risk of confusion when you tell people to write an essentially broken piece of code that you fix up later on. Might consider putting the module creation code _first_ in this document, and then finally have a small segment at the bottom of how it gets added to the index.ts.

File: docs/features/software-catalog/external-integrations.md [hunk 3]
```diff
@@ -1031,3 +1031,360 @@ backend.add(catalogModuleCustomDataParser);
...
+backend.add(catalogModuleCustomIncrementalIngestionProvider);
+
+backend.start();
+```
+
+### Writing an Incremental Entity Provider
+
+To create an Incremental Entity Provider, you need to know how to retrieve a single page of the data that you wish to ingest into the Backstage catalog. If the API has pagination and you know how to make a paginated request to that API, you'll be able to implement an Incremental Entity Provider for this API. For more information about compatibility, check out the [requirements](#requirements) section of this page.
+
+Here is the type definition for an Incremental Entity Provider.
```
Comment (freben) [inline docs/features/software-catalog/external-integrations.md:1107]: Mmm maybe don't put the full source code of that in here. It'll just turn out to drift from reality when we iterate on code. Sometimes it's better to have a little bit of plain text saying something like "Incremental providers need to have a unique name, and some methods that move the cursor and fetch a page of data." and then when the actual implementation starts below, it'll be clear what you are doing in those terms.

File: docs/features/software-catalog/external-integrations.md [hunk 4]
```diff
@@ -1031,3 +1031,360 @@ backend.add(catalogModuleCustomDataParser);
...
+          });
+        },
+      });
+    },
+  });
+```
+
+That's it!!!
+
+### Administrative Routes
```
Comment (freben) [inline docs/features/software-catalog/external-integrations.md:1369]: same here, not sure anyone is meant to actually interface with these, could be in the module's readme instead perhaps?

This document could still have some link like "For a deep dive in the technicalities of the incremental readme, see `<LINK>`"

**Review by freben — COMMENTED** [review]
Alright, thanks, some initial thoughts

**Comment by zeshanziya** [comment]
Thanks @freben for quick review. I will work on suggestions and update.

**Comment by zeshanziya** [comment]
Hey @freben, I have worked on suggestions. Please review.

File: docs/features/software-catalog/external-integrations.md [hunk 5]
```diff
@@ -1031,3 +1031,288 @@ backend.add(catalogModuleCustomDataParser);
...
+backend.start();
+```
+
+### Writing an Incremental Entity Provider
+
+To create an Incremental Entity Provider, you need to know how to retrieve a single page of data from an API with pagination. The `IncrementalEntityProvider` facilitates this by requiring:
+
+- **getProviderName:** A unique name to avoid conflicts with other providers.
+- **next:** Fetches a specific page of entities, moving the cursor forward.
+- **around:** Handles setup and teardown, wrapping the process that iterates through multiple pages.
```
Comment (freben) [inline docs/features/software-catalog/external-integrations.md:1067]:
```suggestion
- **around:** Handles setup and tear-down, wrapping the process that iterates through multiple pages.
```

File: docs/features/software-catalog/external-integrations.md [hunk 6]
```diff
@@ -1031,3 +1031,288 @@ backend.add(catalogModuleCustomDataParser);
...
+}
+```
+
+If you need to pass a token to your API, then you can create a constructor that will receive a token and use the token to setup the client.
+
+```ts
+export class MyIncrementalEntityProvider
+  implements IncrementalEntityProvider<Cursor, Context>
+{
+  token: string;
```
Comment (freben) [inline docs/features/software-catalog/external-integrations.md:1139]:
```suggestion
  private readonly token: string;
```

File: docs/features/software-catalog/external-integrations.md [hunk 7]
```diff
@@ -1031,3 +1031,288 @@ backend.add(catalogModuleCustomDataParser);
...
+  }
+}
+```
+
+The last step is to implement the actual `next` method that will accept the cursor, call the API, process the result and return the result.
+
+```ts
+export class MyIncrementalEntityProvider implements IncrementalEntityProvider<Cursor, Context> {
+
+  token: string;
```
Comment (freben) [inline docs/features/software-catalog/external-integrations.md:1162]:
```suggestion
  private readonly token: string;
```

File: docs/features/software-catalog/external-integrations.md [hunk 8]
```diff
@@ -1031,3 +1031,288 @@ backend.add(catalogModuleCustomDataParser);
...
+    this.token = token;
+  }
+
+  getProviderName() {
+    return `MyIncrementalEntityProvider`;
+  }
+
+
+  async around(burst: (context: Context) => Promise<void>): Promise<void> {
+
```
Comment (freben) [inline docs/features/software-catalog/external-integrations.md:1174]:
```suggestion
```

File: docs/features/software-catalog/external-integrations.md [hunk 9]
```diff
@@ -1031,3 +1031,288 @@ backend.add(catalogModuleCustomDataParser);
...
+        apiVersion: 'backstage.io/v1beta1',
+        kind: 'Component',
+        metadata: {
+          name: item.name,
+          annotations: {
+            // You need to define these, otherwise they'll fail validation
+            [ANNOTATION_LOCATION]: this.getProviderName(),
+            [ANNOTATION_ORIGIN_LOCATION]: this.getProviderName(),
+          }
+        }
```
Comment (freben) [inline docs/features/software-catalog/external-integrations.md:1204]:
```suggestion
        },
```

File: docs/features/software-catalog/external-integrations.md [hunk 10]
```diff
@@ -1031,3 +1031,288 @@ backend.add(catalogModuleCustomDataParser);
...
+        spec: {
+          type: 'service'
```
Comment (freben) [inline docs/features/software-catalog/external-integrations.md:1206]:
```suggestion
          type: 'service',
          lifecycle: 'production', // Ideally your source has this information
          owner: 'unknown', // Ideally your source has this information
```

these are required fields, might as well make the example compile and ingest fully successfully

**Review by freben — APPROVED** [review]
Nice. Gonna self-merge some small tweaks and then ship

**Pull request merged**